### ISSUE TITLE
Do success codes get lost now?

### ISSUE THREAD

**Issue by knopp** [microsoft/windows-rs#922]
For example `DoDragDrop` returns `DRAGDROP_S_DROP` on successful drop. However there doesn't seem to be any way to retrieve this value, since windows-rs calls `ok()` on the `HRESULT`. And given that `DRAGDROP_S_DROP` doesn't have the error bit set it is not passed as error in the `Result`.

Is this an oversight?

**Comment by PhoNoTheSky** [comment]
If you check [here](https://github.com/microsoft/windows-rs/blob/master/src/result/hresult.rs#L18), and place the value of DRAGDROP_S_DROP(262400) where self.0 is, it returns true. but so does DRAGDROP_S_CANCEL(262401). So you want identify which one you got?

**Comment by knopp** [comment]
That's not really the point. The problem is that you can not get the original hresult, because if positive it never leaves `DoDragDrop()`. This is what generated `DoDragDrop()` code looks like:

```rust
pub unsafe fn DoDragDrop<'a>(
    pdataobj: impl ::windows::IntoParam<'a, IDataObject>,
    pdropsource: impl ::windows::IntoParam<'a, IDropSource>,
    dwokeffects: u32,
    pdweffect: *mut u32,
) -> ::windows::Result<()> {
    #[cfg(windows)]
    {
        #[link(name = "OLE32")]
        extern "system" {
            fn DoDragDrop(
                pdataobj: ::windows::RawPtr,
                pdropsource: ::windows::RawPtr,
                dwokeffects: u32,
                pdweffect: *mut u32,
            ) -> ::windows::HRESULT;
        }
        DoDragDrop(
            pdataobj.into_param().abi(),
            pdropsource.into_param().abi(),
            ::std::mem::transmute(dwokeffects),
            ::std::mem::transmute(pdweffect),
        )
        // Here a positive HRESULT gets swallowed.
        .ok()
    }
    #[cfg(not(windows))]
    unimplemented!("Unsupported target OS");
}
```

**Comment by tim-weis** [comment]
This potential issue [has been acknowledged](https://github.com/microsoft/windows-rs/pull/870#discussion_r654495665). I don't know whether this has been missed in the implementation, or OLE just being odd. Either way it's an issue that would need to be addressed.

**Comment by kennykerr** [comment]
Yes, I'm considering a few options for dealing with this. One is to preserve the original signature as a fallback (#737) or a custom `Result<T>` that retains the `HRESULT` even in the success case. I'd prefer the latter because the former would lead to an explosion of the API surface or two different flavors of the Windows API, which I'd like to avoid.

**Comment by knopp** [comment]
For me custom result object would definitely solve the issue. The object could also implement `Deref/DerefMut` for convenient access to actual value.

**Comment by rylev** [comment]
I'm unsure which way to achieve this is best. `Result` is used _a lot_ in the ecosystem, and people will expect any type named `Result` to behave like `std::result::Result` does (with its many helper methods). 

Two variants I see: 
* Methods return `std::result::Result<SomeWrapper<T>, window::Error>`. The `SomeWrapper` derefs to the inner `T` so the inner `T`'s methods can still be used. Passing this value to methods which expect `T` will require a call to some method like `into()` or using deref doing `*wrapper`. 
* Methods return `windows::Result<T>` which derefs to `std::result::Result<T, windows::Error>` but also contains the success `HRESULT`. This makes pattern matching more difficult though as the user would either have to deref the value to cast it to `std::result::Result` or use whatever structure `windows::Result<T>` uses.

**Comment by damyanp** [comment]
The vast majority of APIs behave nicely and only ever return S_OK on success.  For those that return S_FALSE or any of the other non-error result codes, we should get those tagged in the metadata so that projections can behave differently.  I don't think that this problem will be unique to rust.

**Comment by kennykerr** [comment]
I love @damyanp's suggestion. He's absolutely right of course. The vast majority of APIs don't need this and penalizing every API to accommodate the handful that use odd success codes doesn't make sense. Transferring to the win32 metadata repo - hopefully they can add an attribute to such functions/methods and then languages like Rust can simply project their signatures accordingly.
